File: backend/infrastructure/persistence/auth_operation.go
package persistence

import (
	"context"
	"fmt"
	"log/slog"

	"github.com/kanaru-ssk/payment-manager/backend/domain/auth"
	"github.com/kanaru-ssk/payment-manager/backend/domain/emailaddress"
	"google.golang.org/api/identitytoolkit/v3"
)

type operation struct {
	idp         *identitytoolkit.Service
	frontendUrl string
}

func NewAuthOperation(
	idp *identitytoolkit.Service,
	frontendUrl string,
) auth.Operation {
	return &operation{idp: idp, frontendUrl: frontendUrl}
}

func (o *operation) SendSignInLink(ctx context.Context, email emailaddress.EmailAddress) error {
	req := &identitytoolkit.Relyingparty{
		Email:       email.String(),
		RequestType: "EMAIL_SIGNIN",
		ContinueUrl: fmt.Sprintf("%s/signin/complete", o.frontendUrl),
	}
	res, err := o.idp.Relyingparty.GetOobConfirmationCode(req).Do()
	slog.InfoContext(ctx, "persistence.operation.SendSignInLink o.idp.Relyingparty.GetOobConfirmationCode", slog.Any("res", res))

	return err
}

func (o *operation) SignInWithLink(ctx context.Context, email emailaddress.EmailAddress, verificationToken string) (*auth.Auth, error) {
	req := &identitytoolkit.IdentitytoolkitRelyingpartyEmailLinkSigninRequest{
		Email:   email.String(),
		OobCode: verificationToken,
	}
	res, err := o.idp.Relyingparty.EmailLinkSignin(req).Do()
	fmt.Println("debug: res:", res)
	if err != nil {
		return nil, err
	}
	e, err := emailaddress.NewEmailAddress(res.Email)
	if err != nil {
		return nil, err
	}

	return &auth.Auth{
		UserId:       res.LocalId,
		Email:        e,
		IdToken:      res.IdToken,
		RefreshToken: res.RefreshToken,
	}, nil
}

File: frontend/src/infrastructure/persistence/auth-operation.ts
import "server-only";
import type { SendSigninLink, SignInWithLink } from "@/domain/auth";
import { STATUS, getStatusByCode } from "@/domain/status";
import { authServiceClient } from "@/infrastructure/appgrpc/client";

export const sendSigninLink: SendSigninLink = async (email) => {
	try {
		const response = await authServiceClient.sendSignInLink({ email });
		console.log(response);
		const status = getStatusByCode(response.status?.code);

		// status ok
		if (status.code === 0) {
			return {
				success: true,
				data: status,
			};
		}

		// status error
		return {
			success: false,
			errors: status,
		};
	} catch (err) {
		console.log(err);
		// unknown error
		return {
			success: false,
			errors: STATUS.ERR_UNKNOWN,
		};
	}
};

export const signInWithLink: SignInWithLink = async (
	email,
	verificationToken,
) => {
	try {
		const response = await authServiceClient.signInWithLink({
			email,
			verificationToken,
		});
		console.log(response);
		const status = getStatusByCode(response.status?.code);

		// status ok
		if (status.code === 0 && response.auth) {
			return {
				success: true,
				data: {
					userId: response.auth.userId,
					userName: response.auth.userName,
					email: response.auth.email,
					idToken: response.auth.idToken,
					refreshToken: response.auth.refreshToken,
				},
			};
		}

		// status error
		return {
			success: false,
			errors: status,
		};
	} catch (err) {
		// unknown error
		return {
			success: false,
			errors: STATUS.ERR_UNKNOWN,
		};
	}
};
